src/libinjector/linux/methods/linux_read_file.c

```diff
@@ -0,0 +1,371 @@
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ * This file was created by Manorit Chawdhry.                              *
+ * It is distributed as part of DRAKVUF under the same license             *
+ ***************************************************************************/
+
+
+#include <libinjector/debug_helpers.h>
+
+#include "linux_read_file.h"
+#include "linux_syscalls.h"
+
+static event_response_t cleanup(drakvuf_t drakvuf, drakvuf_trap_info_t* info, bool clear_trap);
+static bool write_buffer_to_file(drakvuf_t drakvuf, drakvuf_trap_info_t* info, int amount);
+
+bool init_read_file_method(injector_t injector, const char* file)
+{
+    FILE* fp = fopen(file, "wb");
+    if (!fp)
+    {
+        fprintf(stderr, "Could not open (%s) for writing\n", file);
+        return false;
+    }
+
+    injector->buffer.total_processed = 0;
+    injector->buffer.len = 0;
+
+    injector->buffer.data = g_malloc0(FILE_BUF_SIZE);
+
+    PRINT_DEBUG("File init success\n");
+
+    injector->fp = fp;
+
+    return true;
+}
+
+/* This function handles writing file to guest OS, it does that in total of 6 steps
+ *
+ * STEP1:
+ * It finds the location of syscall present inside vdso, it is used for jumping
+ * to syscall instruction, After that it calls mmap, mmap is used as a buffer for exchanging
+ * data between the guest VM and the host OS. The initial trap must not be removed here,
+ * it should be done in the last step as it will be used for cleanup
+ *
+ * STEP2:
+ * Saves the mmap address and opens the file handle inside guest VM
+ *
+ * STEP3:
+ * This step is the initialization before the while loop analogy, it will verify the file descriptor
+ * and sets up read syscall for reading the initial chunk of file from the target file
+ *
+ * STEP4:
+ * We will reach this after the callback from STEP3, In the beginning of this step, we know that
+ * we have the file chunk in injector buffer now. It will write that buffer to file and set up read
+ * syscall again for reading the next chunk
+ * it overrides the next step as STEP4 until the buffer read from target file is 0.
+ * When it reaches zero, it closes the file handle inside the guest OS which tells us that the
+ * read file operation is complete
+ *
+ * STEP5:
+ * It restores the state of the VM
+ *
+ * STEP6:
+ * It removes the initial trap and exits out of drakvuf loop
+ */
+event_response_t handle_read_file(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    injector_t injector = (injector_t)info->trap->data;
+
+    event_response_t event;
+
+    switch (injector->step)
```

Comment:
The steps in this switch should be broken out into standalone functions

Reply:
Yeah, I was thinking this actually. There are some common steps and functions (`cleanup`) in individual files right now which can be combined ig. I would like to refactor this in all the methods. Will it be fine if I do this in the another refactor PR?

Reply:
I would prefer if it's done now because CodeQL is flagging it

Reply:
Ig that flag is just a warning ( switch case being long, as the writefile method also had it but the check seemed to be successful ). The main thing which it seems to be flagging is something related to `fopen` creating files in `0666` permission ig. Am I right?

Reply:
I don't really care much about the file creation flags. The overly long switch is what I would like fixed.

Reply:
I refactored the boilerplate codes with some helper functions. Hope it seems good to you now